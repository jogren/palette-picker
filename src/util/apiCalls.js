export const postNewProject = async (name) => {
  const url = 'https://palette-picker-api-sfjo.herokuapp.com/api/v1/projects';
  let options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ name })
  }

  const response = await fetch(url, options);
  if(!response.ok) {
    throw Error('There was an issue posting your project');
  }

  const project = await response.json();
  console.log(project);
  return project;
}