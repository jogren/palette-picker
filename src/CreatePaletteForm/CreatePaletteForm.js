import React, { Component } from 'react';
import { connect } from 'react-redux';
import { postNewPalette, getSelectedPalettes, editPalette } from '../util/apiCalls';
import { setSelectedPalettes, clearSelectedPaletteId, hasErrored, setCurrentProjectId } from '../actions';
import { bindActionCreators } from 'redux';

export class CreatePaletteForm extends Component {
  constructor() {
    super();
    this.state = {
      name: "",
      currentProject: ""
    }
  }

  handleChange = (e) => {
    this.setState({[e.target.name]: e.target.value});
  }

  handleCurrentProject = (target) => {
    this.setState({currentProject: target})
  }

  handleSubmit = async (e) => {
    e.preventDefault();
    const { currentProjects, currentPalette, setSelectedPalettes, hasErrored, setCurrentProjectId} = this.props;
    const { name, currentProject } = this.state;
    const projectId = currentProjects.find(project => project.name === currentProject).id;
    const postPalette = {
      name: name,
      project_id: projectId,
      color1: currentPalette[0].hexCode,
      color2: currentPalette[1].hexCode,
      color3: currentPalette[2].hexCode,
      color4: currentPalette[3].hexCode,
      color5: currentPalette[4].hexCode
    }
    try {
      await postNewPalette(postPalette);
      const updatePalettes = await getSelectedPalettes(projectId);
      setCurrentProjectId(projectId);
      setSelectedPalettes(updatePalettes);
    } catch ({message}) {
      hasErrored(message);
    }
    this.props.setRandomPalette();
    this.setState({ name: "" });
  }

  handleSaveEdits = async () => {
    const { currentPalette, currentPaletteId, setSelectedPalettes, clearSelectedPaletteId, setRandomPalette, hasErrored } = this.props;
    const paletteToEdit = {
      color1: currentPalette[0].hexCode,
      color2: currentPalette[1].hexCode,
      color3: currentPalette[2].hexCode,
      color4: currentPalette[3].hexCode,
      color5: currentPalette[4].hexCode
    }
    try {
      await editPalette(paletteToEdit, currentPaletteId.id);
      const updatePalettes = await getSelectedPalettes(currentPaletteId.projectId);
      setSelectedPalettes(updatePalettes);
    } catch ({message}) {
      hasErrored(message);
    }
    clearSelectedPaletteId();
    setRandomPalette();
  }

  revertEdits = () => {
    const { clearSelectedPaletteId, setRandomPalette } = this.props;
    clearSelectedPaletteId();
    setRandomPalette();
  }

  render() {
    const { currentProjects, currentPaletteId } = this.props;
    let displayProjects = currentProjects.map(project => {
      return <option key={project.name} value={project.name}>{project.name}</option>
    })
    return (
      <section className="CreatePalette_section">
        {currentPaletteId && <div className="div_selected-palette">
          <button onClick={this.handleSaveEdits}>{`Save Changes for ${currentPaletteId.name}`}</button>
          <button onClick={this.revertEdits}>Go Back</button>
        </div>}
        {!currentPaletteId && <form className="CreatePalette_form">
          <select onChange={(e) => this.handleCurrentProject(e.target.value)} >
            <option value="">Pick Project</option>
            {displayProjects}
          </select>
          <input
            type="text"
            placeholder="Palette Name..."
            name="name"
            value={this.state.name}
            onChange={this.handleChange} 
            autoComplete="off"/>
          <button disabled={!this.state.name || !this.state.currentProject} onClick={this.handleSubmit}>Save</button>
        </form>}
      </section>
    )
  }
}

export const mapStateToProps = ({ currentProjects, currentPalette, currentPaletteId }) => ({
  currentProjects,
  currentPalette,
  currentPaletteId
});

export const mapDispatchToProps = dispatch => (
  bindActionCreators({ setSelectedPalettes, clearSelectedPaletteId, hasErrored, setCurrentProjectId }, dispatch)
);

export default connect(mapStateToProps, mapDispatchToProps)(CreatePaletteForm);