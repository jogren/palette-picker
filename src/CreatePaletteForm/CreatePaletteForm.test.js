import React from 'react';
import { shallow } from 'enzyme';
import { CreatePaletteForm } from './CreatePaletteForm';
import { postNewPalette, getSelectedPalettes, editPalette } from '../util/apiCalls';
import { mapStateToProps, mapDispatchToProps } from './CreatePaletteForm';

jest.mock('../util/apiCalls');

describe('CreatePaletteForm', () => {
  let wrapper;
  let mockSetSelectedPalettes = jest.fn();
  let mockClearSelectedPaletteId = jest.fn();
  let mockHasErrored = jest.fn();
  let mockSetRandomPalette = jest.fn()
  const currentProjectsMock = [{
    id: 3,
    name: "Winter",
    created_at: "2019-10-10T19:13:08.873Z",
    updated_at: "2019-10-10T19:13:08.873Z",
  }]
  const currentPaletteMock = [
  {
    hexCode: "#25b25d",
    isLocked: false
  },
  {
    hexCode: "#25b25d",
    isLocked: false
  },
  {
    hexCode: "#25b25d",
    isLocked: false
  },
  {
    hexCode: "#25b25d",
    isLocked: false
  },
  {
    hexCode: "#25b25d",
    isLocked: false
  }
]

  beforeEach(() => {
    wrapper = shallow(<CreatePaletteForm 
      currentProjects={currentProjectsMock}
      currentPalette={currentPaletteMock}
      setSelectedPalettes={mockSetSelectedPalettes}
      clearSelectedPaletteId={mockClearSelectedPaletteId}
      hasErrored={mockHasErrored}
      setRandomPalette={mockSetRandomPalette}
      />)
  });

  it('should match the snapshot with the data passed through', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update name state onChange of input', () => {
    wrapper.find('input').at(0).simulate('change', { target: { value: 'New Palette', name: 'name' } });
    expect(wrapper.state('name')).toEqual('New Palette')
  })

  it('should update currentProject state onChange of input', () => {
    wrapper.find('select').at(0).simulate('change', { target: { value: 'Project 1', name: 'currentProject' } });
    expect(wrapper.state('currentProject')).toEqual('Project 1')
  })

  it('should call clearSelectedPaletteId and setRandomPalette when reverEdits is called', () => {
    wrapper.instance().revertEdits();

    expect(mockClearSelectedPaletteId).toHaveBeenCalled();
    expect(mockSetRandomPalette).toHaveBeenCalled();
  })

  it('should run all items in the try when handleSubmit is called', async () => {
    let mockEvent = { preventDefault: () => {} };
    wrapper.setState({ currentProject: 'Winter' });

    wrapper.instance().handleSubmit(mockEvent);
    
    await expect(postNewPalette).toHaveBeenCalled();
    await expect(getSelectedPalettes).toHaveBeenCalled();
    expect(mockSetSelectedPalettes).toHaveBeenCalled();
    expect(wrapper.state('name')).toEqual("")
  })

  it('should run all items in the try when handleSaveEdits is called', async () => {
    wrapper.instance().handleSaveEdits();

    await expect(getSelectedPalettes).toHaveBeenCalled();
    await expect(mockSetSelectedPalettes).toHaveBeenCalled();
    expect(mockClearSelectedPaletteId).toHaveBeenCalled();
    expect(mockSetRandomPalette).toHaveBeenCalled();
  })

  it.skip('should call handleSaveEdits on button click', () => {
    wrapper.instance().handleSaveEdits = jest.fn();
    // wrapper.instance().forceUpdate();
    wrapper.find('button').first().simulate('click');
    expect(wrapper.instance().handleSaveEdits).toHaveBeenCalled()
  })
});