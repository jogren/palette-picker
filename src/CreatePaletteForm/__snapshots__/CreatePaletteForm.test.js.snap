// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePaletteForm should match the snapshot with the data passed through 1`] = `
<section
  className="CreatePalette_section"
>
  <form
    className="CreatePalette_form"
  >
    <select
      onChange={[Function]}
    >
      <option
        value=""
      >
        Pick Project
      </option>
      <option
        key="Winter"
        value="Winter"
      >
        Winter
      </option>
    </select>
    <input
      autoComplete="off"
      name="name"
      onChange={[Function]}
      placeholder="Palette Name..."
      type="text"
      value=""
    />
    <button
      disabled={true}
      onClick={[Function]}
    >
      Save
    </button>
  </form>
</section>
`;
